stylix/palette.nix

```diff
@@ -1,25 +1,106 @@
+  adjustLightness = rgbColorString: primaryScale: let
+    stripped = builtins.replaceStrings ["rgb(" ")"] ["" ""] rgbColorString;
+    values = builtins.split "," stripped;
+    v1 = builtins.fromJSON (builtins.elemAt values 0);
+    v2 = builtins.fromJSON (builtins.elemAt values 2);
+    v3 = builtins.fromJSON (builtins.elemAt values 4);
+    preLightness = (v1 + v2 + v3) / 3.0;
+    adj = (preLightness / 255.0 * (1.0 - primaryScale) + primaryScale) / preLightness * 255.0;
+    v1adj = lib.max (lib.min (v1 * adj) 255.0) 0.0;
+    v2adj = lib.max (lib.min (v2 * adj) 255.0) 0.0;
+    v3adj = lib.max (lib.min (v3 * adj) 255.0) 0.0;
+    round = x: let
+      floored = builtins.floor x;
+      diff = x - floored;
+    in
+      if diff >= 0.5
+      then floored + 1
+      else floored;
+  in
+    builtins.toString (lib.strings.fixedWidthString 2 "0" (lib.toHexString (round v1adj)))
+    + builtins.toString (lib.strings.fixedWidthString 2 "0" (lib.toHexString (round v2adj)))
+    + builtins.toString (lib.strings.fixedWidthString 2 "0" (lib.toHexString (round v3adj)));
+in {
   options.stylix = {
+    scheme = lib.mkOption {
+      type = lib.types.enum [
+        "scheme-content"
+        "scheme-expressive"
+        "scheme-fidelity"
+        "scheme-fruit-salad"
+        "scheme-monochrome"
+        "scheme-neutral"
+        "scheme-rainbow"
+        "scheme-tonal-spot"
+      ];
+      default = "scheme-tonal-spot";
+      description = ''
+        Use this option to select a color scheme type.
+
+        By default we will select matugen's default color scheme type.
+      '';
+    };
+
+    contrast = lib.mkOption {
+      type = lib.types.float;
+      default = 0.0;
+      description = ''
+        Use this option to change the generated color scheme's contrast.
+
+        Value from -1 to 1. -1 represents minimum contrast,
+        0 represents standard (i.e. the design as spec'd),
+        and 1 represents maximum contrast.
+
+        By default, 0 will be used.
+      '';
+    };
+
+    primaryScale = {
+      dark = lib.mkOption {
+        type = lib.types.float;
+        default = 0.0;
+        description = ''
+          Use this option to change the generated dark color scheme's contrast.
+
+          Value from -1 to 1. -1 represents minimum contrast,
+          0 represents standard (i.e. the design as spec'd),
+          and 1 represents maximum contrast.
+
+          By default, 0 will be used.
+        '';
+      };
+      light = lib.mkOption {
+        type = lib.types.float;
+        default = 0.0;
+        description = ''
+          Use this option to change the generated light color scheme's contrast.
+
+          Value from -1 to 1. -1 represents minimum contrast,
+          0 represents standard (i.e. the design as spec'd),
+          and 1 represents maximum contrast.
+
+          By default, 0 will be used.
+        '';
+      };
+    };
+
     polarity = lib.mkOption {
       type = lib.types.enum [
-        "either"
```

Comment:
Consider using one of `lib.types.coercedTo`, `lib.mkChangedOptionModule`, `lib.mkRemovedOptionModule`, etc to handle changes to the polarity option.